src/slptokens/token.h

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2019 Greg Griffith
+// Copyright (c) 2019 The Bitcoin Unlimited developer
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_SLP_TOKEN_H
+#define BITCOIN_SLP_TOKEN_H
+
+#include "script/script.h"
+#include "serialize.h"
+
+#include <stdint.h>
+#include <vector>
+
+enum SLP_TX_TYPE : uint8_t
+{
+    SLP_NULL = 0,
+    SLP_GENESIS,
+    SLP_MINT,
+    SLP_SEND,
+    SLP_COMMIT
+};
+
+class CSLPToken
+{
+    // using vectors for the variable length fields is not ideal, but it is the easiest way to do it
+private:
+    // SLP Fields
+    uint16_t token_type; // 1-2 bytes
+    SLP_TX_TYPE tx_type; // 4-6 bytes
+
+    // NOT IN GENESIS
+    std::vector<uint8_t> token_id; // 32 bytes, token genesis transaction hash
+
+    // NOT IN SEND
+    uint8_t mint_baton_vout; //(0 bytes or 1 byte between 0x02-0xff), if 0 there is no mint baton
+
+    // ONLY IN SEND
+    std::vector<uint64_t> token_output_quantitys; // this can be a max size of 19
+
+    // ONLY IN GENESIS
+    std::vector<uint8_t> token_ticker; // (0 or more bytes, suggested utf-8)
+    std::vector<uint8_t> token_name; // (0 or more bytes, suggested utf-8)
+    std::vector<uint8_t> token_document_url; // (0 or more bytes, suggested ascii)
+    std::vector<uint8_t> token_document_hash; // 32 bytes
+    uint8_t decimals;
+    uint64_t initial_token_mint_quantity;
+    // ONLY IN SEND
+    uint64_t additional_token_quantity;
+
+private:
+    bool DetermineDynamicSize(const CScript &scriptPubKeyIn, size_t &index, uint32_t &return_size);
+    SLP_TX_TYPE ParseType(const CScript &scriptPubKeyIn, size_t &index, const uint32_t &tx_type_size);
+    uint8_t ParseBytesGenesis(const CScript &scriptPubKeyIn, size_t &index);
+    uint8_t ParseBytesMint(const CScript &scriptPubKeyIn, size_t &index);
+    uint8_t ParseBytesSend(const CScript &scriptPubKeyIn, size_t &index);
+    uint8_t ParseBytesCommit(const CScript &scriptPubKeyIn, size_t &index);
+    uint8_t _ParseBytes(const CScript &scriptPubKeyIn);
+
+public:
+    CSLPToken() { SetNull(); }
+    void SetNull()
+    {
+        token_type = 0;
+        tx_type = SLP_NULL;
+        token_id.clear();
+        mint_baton_vout = 0;
+        token_output_quantitys = std::vector<uint64_t>(19, 0); // prefill with 19 0's
+        token_ticker.clear();
+        token_name.clear();
+        token_document_url.clear();
+        token_document_hash.clear();
+        decimals = 0;
+        initial_token_mint_quantity = 0;
+        additional_token_quantity = 0;
+    }
+
+    bool IsNull() { return tx_type == SLP_NULL; }
+    ADD_SERIALIZE_METHODS;
+    // this is not that great, it might be better to make a token header
+    // and then different types that extend the header to avoid always having all fields
+    // in 1 token class like this
```

Comment:
nit: can you clarify this commet a bit...is this a TODO:?  what do you mean when you say, "this is not great"? or "like this"?...I don't understand what you're referring to.

Reply:
It was just a thought i had that maybe should be considered at some later point.

As implemented in this PR the CSLPToken class handles all tokens (gensis, mint, send, commit) and a majority of the data members are only ever used by 1 type. I was thinking it might be better later on to break up the class and use different ones for different types so everything isnt shoved into one class.
Functionally it doesnt matter, would just be a readability or organizational thing